Import PEM keys

vcsjones opened the pull request:
Closes #31201

vcsjones committed:
Import PEM encoded keys.

vcsjones committed:
Documentation.

vcsjones committed:
Fix reference source

vcsjones committed:
Add missing doc comment.

vcsjones committed:
Skip overload tests for shim helpers.

vcsjones committed:
Skip explicit curve test for unsupported platforms

vcsjones commented:
Clearly I boondoggled that last commit. Will fix in the AM.

vcsjones committed:
Remove explicit params test.

This is not strictly related to processing PEMs since the underlying
import methods already test this.

vcsjones commented:
Windows failure is #34045

vcsjones commented:
@bartonjs this is ready for review when ever you have time to spare.

vcsjones committed:
Merge remote-tracking branch 'ms/master' into 31201-impl

bartonjs commented on src/libraries/Common/src/Internal/Cryptography/PemKeyImportHelpers.cs:
```diff
@@ -0,0 +1,157 @@
+                }
+
+                Index offset = fields.Location.End;
+                pem = pem[offset..];
```
```suggestion
                pem = pem.Slice(offset);
```

bartonjs commented on src/libraries/Common/src/Internal/Cryptography/PemKeyImportHelpers.cs:
```diff
@@ -0,0 +1,157 @@
+                }
+
+                Debug.Assert(bytesWritten == base64size);
+                Span<byte> decodedBase64 = decodeBuffer.AsSpan(0, base64size);
```
In bizarro-world, where TryFromBase64Chars reports a different answer in a release build, let's use the answer it gave.

```suggestion
                Span<byte> decodedBase64 = decodeBuffer.AsSpan(0, bytesWritten);
```

bartonjs commented on src/libraries/Common/src/Internal/Cryptography/PemKeyImportHelpers.cs:
```diff
@@ -0,0 +1,157 @@
+                    foundFields = fields;
+                    foundSlice = pem;
+                }
+                else if (label.SequenceEqual("ENCRYPTED PRIVATE KEY"))
```
I don't know if we need a PemHeaders constants class; but at least using a const string in this class would be better than repeating the literal across methods.

bartonjs commented on src/libraries/Common/src/Internal/Cryptography/PemKeyImportHelpers.cs:
```diff
@@ -0,0 +1,157 @@
+            // throw.
+            if (containsEncryptedPem)
+            {
+                throw new ArgumentException(SR.Argument_PemImport_EncryptedPem, nameof(input));
```
Encrypted followed by plain reports ambiguous, plain followed by encrypted reports encrypted.

It seems like either this branch, or the match on the label, should notice that `importAction` is not null, and report ambiguous.

bartonjs commented on src/libraries/Common/src/Internal/Cryptography/PemKeyImportHelpers.cs:
```diff
@@ -0,0 +1,157 @@
+
+                Debug.Assert(bytesWritten == base64size);
+                Span<byte> decodedBase64 = decodeBuffer.AsSpan(0, base64size);
+                importAction(decodedBase64, out _);
```
I assume that you've made a conscious choice to say it's OK to ignore trailing bytes here, since there's no check that the import action reports bytesRead == base64Size.

I can accept that... but a comment saying so would be nice.  (Reasonable justification includes that the key has already been modified, so throwing and not have the same object modification effect)

bartonjs commented on src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/EC/ECKeyPemTests.cs:
```diff
@@ -0,0 +1,255 @@
+        [Fact]
+        public void ImportFromPem_Pkcs8_Simple()
+        {
+            using TAlg key = CreateKey();
```
Please use the braced using statement anywhere a using statement is used.

bartonjs commented on src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/EC/ECKeyPemTests.cs:
```diff
@@ -0,0 +1,255 @@
+Isuk92Ner/JmgKjYoSumHVmSNfZ9nLTVjxeD08pD548KWrqmJAeZNsDDqQ==
+-----END PUBLIC KEY-----";
+            using TAlg key = CreateKey();
+            AssertExtensions.Throws<ArgumentException>("input", () => key.ImportFromPem(pem));
```
I don't see a test that says what happens to this input when sent to the encrypted PEM import.  (Either success, or still throwing). (Perhaps I'm just bad at searching/remembering)

bartonjs commented on src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/RSA/RSAKeyPemTests.cs:
```diff
@@ -0,0 +1,384 @@
+Q079SlHzXuvocqIdt4IM1EmIlrlU9GGvAh8Ijv3FFPUSLfANgfOIH9mX7ldpzzGk
+rmaUzxQvyuVLAiEArCTM8dSbopUADWnD4jArhU50UhWAIaM6ZrKqC8k0RKsCIQDC
+yZWUxoxAdjfrBGsx+U6BHM0Myqqe7fY7hjWzj4aBCw==
+-----END RSA PRIVATE KEY-----";
```
Is it ambiguous on an RSA key for `EC PRIVATE KEY` and `RSA PRIVATE KEY`? What about `EC PRIVATE KEY` and `PRIVATE KEY` (whether or not the PKCS#8 has the RSA Oid)?

I'm not sure I actually care what the answer is, but I'd like to see a "not contextually applicable" format test that used a known-as-a-key-type format.

vcsjones replied:
We touched a bit on this in the original issue, the gist is:

1. For PKCS8/SPKI keys, we don't decode them for purposes of reporting a duplicate. If a PKCS1 and PKCS8 PEM are found, we assume dupe, even in the PKCS8 turns out to be a key for an entirely different algorithm.
2. If the labels are for two known-to-be-not-ambiguous keys, like EC PRIVATE KEY and RSA PRIVATE KEY, we won't report a duplicate. I think the rationale there was that we don't have a crystal ball for future algorithms. If we throw ambiguous for EC PRIVATE KEY for example while importing RSA, and eventually .NET gets enlightened with a new algorithm with PEM encoding (say ED25519 since it doesn't use the regular "EC" label), we wouldn't want to start breaking people's existing PEM aggregates that contained those algorithms we once ignored.

bartonjs replied:
Yep, those both sound reasonable (and what I felt the conclusions had been); I'm after the test that proves it for "known key types" vs "this algorithm's known key types" :smile:

vcsjones replied:
Regardless I will illustrate this better with tests.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/Resources/Strings.resx:
```diff
@@ -81,6 +81,15 @@
   <data name="ArgumentNull_Buffer" xml:space="preserve">
     <value>Buffer cannot be null.</value>
   </data>
+  <data name="Argument_PemImport_NoPemFound" xml:space="preserve">
+    <value>A supported key was not found.</value>
+  </data>
+  <data name="Argument_PemImport_AmbiguousPem" xml:space="preserve">
+    <value>Multiple supported keys were found.</value>
+  </data>
+  <data name="Argument_PemImport_EncryptedPem" xml:space="preserve">
+    <value>An encrypted key was found, but attempted to import it without a password.</value>
+  </data>
```
Some user friendliness enhancement suggestions:

```suggestion
  <data name="Argument_PemImport_NoPemFound" xml:space="preserve">
    <value>No supported key formats were found. Check that the input represents the contents of a PEM-encoded key file, not the path to such a file.</value>
  </data>
  <data name="Argument_PemImport_AmbiguousPem" xml:space="preserve">
    <value>The input contains multiple keys, but only one key can be imported.</value>
  </data>
  <data name="Argument_PemImport_EncryptedPem" xml:space="preserve">
    <value>An encrypted key was found, but no password was provided. Use ImportFromEncryptedPem to import this key.</value>
  </data>
```

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -1125,5 +1125,168 @@ public int GetMaxSignatureSize(DSASignatureFormat signatureFormat)
+        /// </summary>
+        /// <param name="input">The PEM text of the key to import.</param>
+        /// <exception cref="ArgumentException">
+        /// <paramref name="input"/> does not contain a PEM-encoded key with a recognized label.
```
I believe once one para is being used everything needs to be para-ified.

```
<exception cref="ArgumentException">
  <para><paramref name="input"/> does not contain a PEM-encoded key with a recognized label.</para>
  <para>-or</para>
  ...
</exception>
```

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -1125,5 +1125,168 @@ public int GetMaxSignatureSize(DSASignatureFormat signatureFormat)
+        /// <remarks>
+        ///   <para>
+        ///   Unsupported or malformed PEM-encoded objects will be ignored. If multiple supported PEM labels
+        ///   are found, an exception is raised to prevent importing a key when
```
```suggestion
        ///   are found, an exception is thrown to prevent importing a key when
```

vcsjones replied:
It seems it is going to take a while longer for Ruby part of my career path to wear off.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -1125,5 +1125,168 @@ public int GetMaxSignatureSize(DSASignatureFormat signatureFormat)
+        ///   When the base-64 decoded contents of <paramref name="input" /> indicate an algorithm that uses PBKDF1
+        ///   (Password-Based Key Derivation Function 1) or PBKDF2 (Password-Based Key Derivation Function 2),
+        ///   the password is converted to bytes via the UTF-8 encoding. This method
+        ///   only supports the binary (BER/CER/DER) encoding of EncryptedPrivateKeyInfo.
```
I think that this method only supports the PEM encoded version of EncryptedPrivateKeyInfo :smile:.

vcsjones replied:
🤦‍♂

vcsjones replied:
Oh, I recall what I was trying to communicate - it just came out poorly. I was trying to say that the base-64 contents, when decoded, must be a BER, DER, CER {Encrypted}PrivateKeyInfo.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECDiffieHellman.cs:
```diff
@@ -433,5 +434,173 @@ public virtual unsafe bool TryExportECPrivateKey(Span<byte> destination, out int
+                {
+                    return ImportPkcs8PrivateKey;
+                }
+                else if (label.SequenceEqual("PUBLIC KEY"))
```
Since these literals are spread out across multiple files, I think we should go ahead and create a PemConstants/PemHeaders/PemLabels constants file.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSA.cs:
```diff
@@ -1125,5 +1125,168 @@ public int GetMaxSignatureSize(DSASignatureFormat signatureFormat)
+        /// </remarks>
+        public override void ImportFromEncryptedPem(ReadOnlySpan<char> input, ReadOnlySpan<byte> passwordBytes)
+        {
+            PemKeyImportHelpers.ImportEncryptedPem<byte>(input, passwordBytes, ImportEncryptedPkcs8PrivateKey);
```
Since ImportEncryptedPkcs8PrivateKey comes from AsymmetricAlgorithm, why do all of the key types need to implement this method?  Seems like it could just be the virtual implementation...

vcsjones replied:
@bartonjs that's a good point. [The API proposal][1] had all of them overriden. If we remove the `override`s from the RSA/DSA/ECDsa, do we need to go through another round of API Review?

[1]: https://github.com/dotnet/runtime/issues/31201#issue-558482299

bartonjs replied:
Nah, from the pure API perspective it was only the `virtual` that was interesting.  Removing overrides that don't yet actually exist is just an implementation detail.

bartonjs replied:
I'd probably just cut the sentence.  Assuming it's mirroring what I wrote on ImportPkcs8PrivateKey (assuming I wrote something like that), I was just saying "this doesn't understand PEM" :)

bartonjs commented on src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/AsymmetricAlgorithm.cs:
```diff
@@ -164,6 +164,57 @@ protected virtual void Dispose(bool disposing)
+        /// default behavior will throw <see cref="NotImplementedException" />.
+        /// </remarks>
+        public virtual void ImportFromPem(ReadOnlySpan<char> input) =>
+            throw new NotImplementedException(SR.NotSupported_SubclassOverride);
```
On the one hand, it seems like we could have the base class version support SPKI and PKCS8; but since everything except DSA has defined their own specific private key format as a top-level item, I agree that everything would probably end up overriding it anyways.

The encrypted ones seem less convincing... I think we can provide default implementations of those.

vcsjones committed:
Braced usings.

vcsjones committed:
Braced usings.

vcsjones committed:
Braced usings.

vcsjones committed:
Clearer exception message.

vcsjones committed:
Merge remote-tracking branch 'ms/master' into 31201-impl

vcsjones committed:
Use size based on actual decoding.

vcsjones committed:
Extract PEM labels out to constants.

vcsjones committed:
Documentation feedback.

Use para as required.
raised -> thrown.
Remove confusing text around BER,CER,DER.

vcsjones committed:
Report consistent exception regardless of PEM order

vcsjones committed:
Improve tests.

Ensure ArgumentException is thrown with the right reason.

Improve ambiguous PEM tests.

vcsjones committed:
Explain bytesRead discard.

vcsjones committed:
Missing EncryptedPrivateKeyInfo tests.

vcsjones commented:
@bartonjs I think this addresses all of your feedback. The mono/linux/X509Certificates failure is unrelated.
